data/updater.py

```diff
@@ -0,0 +1,357 @@
+HJwUSeLMhRvi4jTT+z5QLdYloFRZmDRwNg63csGZRkly9vjrAiMVHMpcJI0eCei/
+XgeKSxoiAmzNuc2J47SF2z7WIsDwHhwRj6tj4dOW3Ye0WIkcTIvHd7UTVX02v+oB
+d5YhABEBAAGJAiUEGAEKAA8FAlfEAKACGwwFCQHhM4AACgkQw+zcBCBPnSn25BAA
+xU7NKi6BokqnloYncvL74fuUpam3LJBwsOkFehuO2D+X9A2blIpbpXtUoWXwRc9V
+Jf7nL/yhMKcOB9m1MHVPtxtN5JzV9p8k2BT04/X6fa09umsJ3hwg/zhXrkGFrMVE
+hfAk6q8a0Y5oJKUOoJhzxqD9ItibxHPkqb+R/GSMfrDIRE0ecfIltDLIRQMlBF1b
+z5WN5Dwv8cikeQrsK6DNvbUUuHAbG8RZYG9QxFdkbehp46bCA8CfINENBzIskckx
+xwlTtVxcDD0Irql7EuIM1bpWdFxBZPlmcDyLrZgCYgSqPOe2mOK9V37jM3fsTqR9
+C9fmr3DDmAm8XrUBL9ORwMTFa6LIUpUoSSzG258h9qPiBySj7b51QbrJ9WXiISsQ
+5X/V8u6Qbp9PJnEUXXfIE9YVspF3+Zrfn0XqqjEND+bWi52cgDvwROSFf4KQnAto
+FdlxHtxlcZgJEFeHLyb370XI4rcNs7zMoD3B06Eyay5oY98w9JFYwrm6bVvNzis7
+CrK2K46QzSUZm1BHTOi0AlnqpYtLaJnMSQCxTmjyRBaJM5DGXs/86y/OYdBcpiYG
+aV0q9EcvDLQnSpa0vlsVM0AfMY+To1RTCMv+TGKMZHHWeV2yUABWK52raBpQHtjs
+7SonBqg04+2E4w1WmZEx1u9QyDXmlaLxnR+YIqilM7g=
+=RbWA
+-----END PGP PUBLIC KEY BLOCK-----
+"""
+
+
+class RequestFailed(Exception):
+    pass
+
+def get_request(url, follow_redirects=True):
+    cmd = ["curl", "-q"]
+    if follow_redirects is True:
+        cmd.append("-L")
+    cmd.append(url)
+
+    tmp_file = tempfile.TemporaryFile()
+
+    try:
+        check_call(cmd, stdout=tmp_file)
+    except CalledProcessError:
+        raise RequestFailed
+
+    tmp_file.seek(0)
+
+    return tmp_file.read()
+
+def get_current_version():
+    if not os.path.exists(CURRENT_VERSION_PATH):
+        return 0
+    with open(CURRENT_VERSION_PATH) as in_file:
+        version = in_file.read()
+    return int(version)
+
+def get_latest_version():
+    version = get_request(UPDATE_BASE_URL + "latest/version")
+    return int(version)
+
+class InvalidSignature(Exception):
+    pass
+
+class InvalidPublicKey(Exception):
+    pass
+
+
+def verify_file(signature_path, file_path, signer_pk_path):
+    tmp_dir = tempfile.mkdtemp()
+    tmp_key = os.path.join(tmp_dir, "signing-key.gpg")
+
+    try:
+        try:
+            check_call(["gpg", "--batch", "--yes", "-o", tmp_key,
+                        "--dearmor", signer_pk_path])
+        except CalledProcessError:
+            raise InvalidPublicKey
+
+        try:
+            output = check_output(["gpg", "--batch", "--status-fd", "1",
+                                   "--no-default-keyring", "--keyring",
+                                   tmp_key, "--trust-model", "always",
+                                   "--verify", signature_path, file_path])
+        except CalledProcessError:
+            raise InvalidSignature
+
+    except Exception as e:
+        raise e
+
+    finally:
+        shutil.rmtree(tmp_dir)
+
+    return output
+
+class UpdateFailed(Exception):
+    pass
+
+def perform_update(version, skip_verification=False):
+    try:
+        updater = get_request(UPDATE_BASE_URL + "{0}/update.py".format(version))
+        updater_path = os.path.join(UPDATER_PATH, "update-{0}.py".format(version))
```

Comment:
ah, so the scoping rules of python with respect to try...except blocks are that you can define a variable inside such a block and then use it outside!